lib/take_elevated.lua

```diff
@@ -0,0 +1,79 @@
+local tokens_per_ms               = tonumber(ARGV[1])
+local bucket_size                 = tonumber(ARGV[2])
+local tokens_to_take              = tonumber(ARGV[3])
+local ttl                         = tonumber(ARGV[4])
+local drip_interval               = tonumber(ARGV[5])
+local erl_tokens_per_ms           = tonumber(ARGV[6])
+local erl_bucket_size             = tonumber(ARGV[7])
+local erl_activation_period_mins  = tonumber(ARGV[8])
+
+-- check if we should use erl
+local erlKey = KEYS[2]
+local is_erl_activated = redis.call('EXISTS', erlKey)
+
+local current_time = redis.call('TIME')
+local current_timestamp_ms = current_time[1] * 1000 + current_time[2] / 1000
+
+local current = redis.pcall('HMGET', KEYS[1], 'd', 'r')
+
+if current.err ~= nil then
+    current = {}
+end
+
+
+local function calculateNewBucketContent(current, tokens_per_ms, bucket_size, current_timestamp_ms)
+    if current[1] and tokens_per_ms then
+        -- drip bucket
+        local last_drip = current[1]
+        local content = current[2]
+        local delta_ms = math.max(current_timestamp_ms - last_drip, 0)
+        local drip_amount = delta_ms * tokens_per_ms
+        return math.min(content + drip_amount, bucket_size)
+    elseif current[1] and tokens_per_ms == 0 then
+        -- fixed bucket
+        return current[2]
+    else
+        -- first take of the bucket
+        return bucket_size
+    end
+end
+
+local new_content = calculateNewBucketContent(current, tokens_per_ms, bucket_size, current_timestamp_ms)
+local enough_tokens = new_content >= tokens_to_take
+
+-- https://redis.io/docs/interact/programmability/eval-intro/#:~:text=scripts%20debugger.-,Script%20replication,-In%20standalone%20deployments
+redis.replicate_commands()
+
+if enough_tokens and is_erl_activated==0 then
+    new_content = math.min(new_content - tokens_to_take, bucket_size)
+else
+    -- use erl
+    new_content = calculateNewBucketContent(current, erl_tokens_per_ms, erl_bucket_size, current_timestamp_ms)
+
+    -- if activating erl for first time, refill the bucket with the old bucket size
+    if is_erl_activated == 0 then
```

Comment:
Control is done outside of the script, [right here.](https://github.com/ecita/limitd-redis/blob/93406e47e321dfb5189f2fa3a40e4ec5ab3948ba/lib/db.js#L251)
However, we've not yet implemented the "enabled/disabled" logic per bucket yet -- this will be another PR, and it will add `erlEnabled:true|false` to bucket definition under `elevated_limits` section.

Current implementation assumes that ERL can be used if:
1. `allowERL` was passed `true` while calling `take()`,
AND
2. `elevated_limits` config exists in bucket configuration. (see with the link above)